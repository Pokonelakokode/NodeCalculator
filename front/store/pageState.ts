import {createSlice, PayloadAction} from "@reduxjs/toolkit";

enum PageStateActionTypes {
    ADD_TO_CURRENT_VALUE = "ADD_TO_CURRENT_VALUE",
    REMOVE_FROM_CURRENT_VALUE = "REMOVE_FROM_CURRENT_VALUE",
    SET_OBJ = "SET_OBJ",
    ADD_OPERATION = "ADD_OPERATION",
    EXECUTE = "EXECUTE",
}

export interface IOperation {
    operator: Operator;
    value: string;
}

export type Operator = "+" | "-" | "*" | "/";

interface IPageState {
    readonly baseValue: string;
    readonly operations: IOperation[];
    readonly selectedOperator: Operator | null;
    readonly loaded: boolean;
    readonly mem: number;
}

export interface IPageStateActions {
    SET_OBJ: PayloadAction<Partial<IPageState>>;
    ADD_TO_CURRENT_VALUE: PayloadAction<string>;
    REMOVE_FROM_CURRENT_VALUE: PayloadAction;
    ADD_OPERATION: PayloadAction<{operator: Operator}>;
    EXECUTE: PayloadAction;
}

const initialState: IPageState = {
    baseValue: "0",
    loaded: false,
    mem: 0,
    operations: [],
    selectedOperator: null,
};

const pageState = createSlice({
    initialState,
    name: "pageState",
    reducers: {
        [PageStateActionTypes.SET_OBJ]:
            (state, action: IPageStateActions["SET_OBJ"]) => {
                return {...state, ...action.payload};
            },
        [PageStateActionTypes.ADD_TO_CURRENT_VALUE]:
            (state, action: IPageStateActions["ADD_TO_CURRENT_VALUE"]) => {
                return state.operations.length > 0 ?
                    {
                        ...state,
                        operations: state.operations.map((operation, index) => (
                            index < state.operations.length - 1 ?
                                operation :
                                {...operation, value: sanitizeNumber(operation.value + action.payload)}
                        )),
                    } :
                    {
                        ...state,
                        baseValue: sanitizeNumber(state.baseValue + action.payload),
                    };
            },
        [PageStateActionTypes.REMOVE_FROM_CURRENT_VALUE]:
            (state, action: IPageStateActions["REMOVE_FROM_CURRENT_VALUE"]) => {
                return state.operations.length > 0 ?
                    {
                        ...state,
                        operations: state.operations.reduce((agg: IOperation[], operation, index) => {
                            index < state.operations.length - 1 ?
                                agg.push(operation) :
                                // tslint:disable-next-line:no-unused-expression
                                operation.value.length > 1 && agg.push(
                                    {...operation, value: operation.value.slice(0, -1)},
                                );

                            return agg;
                        },[]),
                    } :
                    {
                        ...state,
                        baseValue: state.baseValue.slice(0, -1),
                    };

            },
        [PageStateActionTypes.ADD_OPERATION]:
            (state, action: IPageStateActions["ADD_OPERATION"]) => {
            return {
                ...state,
                operations: [...state.operations, {operator: action.payload.operator, value: "0"}],
            };
        },
        [PageStateActionTypes.EXECUTE]:
            (state, action: IPageStateActions["EXECUTE"]) => {
                return {
                    ...state,
                    baseValue: getCalculation(state.baseValue, state.operations).toString(),
                    operations: [],
                };
            },
    },
});

function sanitizeNumber(str: string): string {
    let occurence = 0;
    return str.replace(/^0+/, "").replace(/\./g,(match) => {
        occurence++;
        return occurence > 1 ? "" : match;
    });
}

export function isOperator(str: string): str is Operator {return ["+", "-", "*", "/"].includes(str)}

export function getCalculation(baseValue: string, operations: IOperation[]) {
    const parsedValue = parseFloat(baseValue);
    const calculation =  operations.reduce((agg: number, operation) => {
        const val = parseFloat(operation.value);
        if(isNaN(val)) { return agg; }
        switch (operation.operator) {
            case "*":
                agg = agg * val;
                break;
            case "/":
                agg = agg / val;
                break;
            case "+":
                agg += val;
                break;
            case "-":
                agg -= val;
                break;
        }
        return agg;
    }, isNaN(parsedValue) ? 0 : parsedValue);
    return isNaN(calculation) ? "Not a number" : calculation;
}

export const pageStateActions = pageState.actions;
export const pageStateReducer = pageState.reducer;

export default pageState;
