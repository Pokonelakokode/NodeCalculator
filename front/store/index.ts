import {combineReducers, configureStore, createSelector} from "@reduxjs/toolkit";
import {fileMem} from "./middlewares/fileMem";
import {IPageStateActions, pageStateReducer} from "./pageState";

export const rootReducer = combineReducers({
    pageState: pageStateReducer,
});

export const store = configureStore<RootState>({
    devTools: process.env.NODE_ENV === "production" ? false : {trace: true, traceLimit: 25},
    middleware: [fileMem],
    reducer: rootReducer,
});

export const mainSelector = (state: RootState) => state.pageState;

export type RootState = ReturnType<typeof rootReducer>;
export type RootActions = IPageStateActions[keyof IPageStateActions];
