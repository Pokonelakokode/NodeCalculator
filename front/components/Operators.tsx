import * as React from "react";
import {useDispatch} from "react-redux";
import {bindActionCreators} from "redux";

import {pageStateActions} from "../store/pageState";
import Button from "./Button";

const Operators: React.FC = (props) => {
    const actions = bindActionCreators({
        addOperation: pageStateActions.ADD_OPERATION,
        execute: pageStateActions.EXECUTE,

        setObj: pageStateActions.SET_OBJ,
    }, useDispatch());
    return (
        <div>
            <div className="row no-gutters">
                <Button label={"+"} col={6} onClick={() => actions.addOperation({operator: "+"})}/>
                <Button label={"-"} col={6} onClick={() => actions.addOperation({operator: "-"})}/>
            </div>
            <div className="row no-gutters">
                <Button label={"*"} col={6} onClick={() => actions.addOperation({operator: "*"})}/>
                <Button label={"/"} col={6} onClick={() => actions.addOperation({operator: "/"})}/>
            </div>
            <div className="row">
                <Button label={"="} col={12} onClick={() => {
                    actions.execute();
                }}/>
            </div>

        </div>
    );
};

export default Operators;
