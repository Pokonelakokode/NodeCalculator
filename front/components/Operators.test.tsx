import React from "react";
import {Provider} from "react-redux";
import {create} from "react-test-renderer";
import {store} from "../store";
import Operators from "./Operators";

describe("Operators component", () => {
    it("should be able to render", () => {
        const component = create(
            <Provider store={store}>
                <Operators />
            </Provider>,
        );
        expect(component.toJSON()).toMatchSnapshot();
    });
});
