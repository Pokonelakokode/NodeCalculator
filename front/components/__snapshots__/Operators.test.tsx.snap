// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Operators component should be able to render 1`] = `
<div>
  <div
    className="row no-gutters"
  >
    <div
      className="col-6 mt-2"
    >
      <div
        className="numeric"
        onClick={[Function]}
      >
        +
      </div>
    </div>
    <div
      className="col-6 mt-2"
    >
      <div
        className="numeric"
        onClick={[Function]}
      >
        -
      </div>
    </div>
  </div>
  <div
    className="row no-gutters"
  >
    <div
      className="col-6 mt-2"
    >
      <div
        className="numeric"
        onClick={[Function]}
      >
        *
      </div>
    </div>
    <div
      className="col-6 mt-2"
    >
      <div
        className="numeric"
        onClick={[Function]}
      >
        /
      </div>
    </div>
  </div>
  <div
    className="row"
  >
    <div
      className="col-12 mt-2"
    >
      <div
        className="numeric"
        onClick={[Function]}
      >
        =
      </div>
    </div>
  </div>
</div>
`;
